lib/spack/spack/relocate.py

```diff
@@ -257,34 +324,65 @@ def needs_text_relocation(filetype):
     return ("text" in filetype)
 
 
-def relocate_binary(path_names, old_dir, new_dir):
+def relocate_binary(path_names, old_dir, new_dir, allow_root):
     """
     Change old_dir to new_dir in RPATHs of elf or mach-o files
+    Account for the case where old_dir is now a placeholder
     """
+    placeholder = set_placeholder(old_dir)
     if platform.system() == 'Darwin':
         for path_name in path_names:
-            rpaths, deps, idpath = macho_get_paths(path_name)
-            new_rpaths, new_deps, new_idpath = macho_replace_paths(old_dir,
-                                                                   new_dir,
-                                                                   rpaths,
-                                                                   deps,
-                                                                   idpath)
+            (rpaths, deps, idpath) = macho_get_paths(path_name)
+            # new style buildaches with placeholder in binaries
+            if (deps[0].startswith(placeholder) or
+                rpaths[0].startswith(placeholder) or
+                (idpath and idpath.startswith(placeholder))):
+                (new_rpaths,
+                 new_deps,
+                 new_idpath) = macho_replace_paths(placeholder,
+                                                   new_dir,
+                                                   rpaths,
+                                                   deps,
+                                                   idpath)
+            # old style buildcaches with original install root in binaries
+            else:
+                (new_rpaths,
+                 new_deps,
+                 new_idpath) = macho_replace_paths(old_dir,
+                                                   new_dir,
+                                                   rpaths,
+                                                   deps,
+                                                   idpath)
             modify_macho_object(path_name,
                                 rpaths, deps, idpath,
                                 new_rpaths, new_deps, new_idpath)
+            if (not allow_root and
+                strings_contains_installroot(path_name, old_dir)):
+                    raise InstallRootStringException(path_name, old_dir)
+
     elif platform.system() == 'Linux':
         for path_name in path_names:
             orig_rpaths = get_existing_elf_rpaths(path_name)
+            # new style buildaches with placeholder in binaries
```

Comment:
IMO it's worth moving this down a couple lines right above `if orig_rpaths[0].startswith(placeholder):` since `if orig_rpaths:` also contains logic for old-style build caches.